File: FE/src/app/services/userService.ts
import api from './api';

export interface UserProfile {
  id: string;
  fullName: string;
  email: string;
  phone?: string;
  roles: string[];
  studentId?: string;
  major?: string;
  year?: string;
  avatarUrl?: string;
  walletBalance: number;
  status: string;
  createdAt: string;
}

export interface UpdateProfilePayload {
  fullName: string;
  phone?: string;
  studentId?: string;
  major?: string;
  year?: string;
  avatarUrl?: string;
}

export async function getMe(): Promise<UserProfile> {
  const { data } = await api.get('/api/users/me');
  return data.data as UserProfile;
}

export async function updateProfile(payload: UpdateProfilePayload): Promise<UserProfile> {
  const { data } = await api.put('/api/users/me', payload);
  return data.data as UserProfile;
}

export async function changePassword(currentPassword: string, newPassword: string): Promise<void> {
  await api.post('/api/users/me/password', { currentPassword, newPassword });
}

File: BE/src/main/java/com/dynforge/be/service/UserService.java
package com.dynforge.be.service;

import com.dynforge.be.exception.BadRequestException;
import com.dynforge.be.mapper.UserMapper;
import com.dynforge.be.model.dto.ChangePasswordRequest;
import com.dynforge.be.model.dto.UpdateProfileRequest;
import com.dynforge.be.model.dto.UserResponse;
import com.dynforge.be.model.entity.User;
import com.dynforge.be.repository.RefreshTokenRepository;
import com.dynforge.be.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.bson.types.ObjectId;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;
    private final RefreshTokenRepository refreshTokenRepository;
    private final PasswordEncoder passwordEncoder;
    private final UserMapper userMapper;

    public UserResponse updateProfile(User user, UpdateProfileRequest request) {
        user.setFullName(request.fullName());
        user.setPhone(request.phone());
        user.setStudentId(request.studentId());
        user.setMajor(request.major());
        user.setYear(request.year());
        if (request.avatarUrl() != null) {
            user.setAvatarUrl(request.avatarUrl());
        }
        return userMapper.toResponse(userRepository.save(user));
    }

    public void changePassword(User user, ChangePasswordRequest request) {
        if (!passwordEncoder.matches(request.currentPassword(), user.getPasswordHash())) {
            throw new BadRequestException("Current password is incorrect");
        }
        if (passwordEncoder.matches(request.newPassword(), user.getPasswordHash())) {
            throw new BadRequestException("New password must be different from the current one");
        }

        user.setPasswordHash(passwordEncoder.encode(request.newPassword()));
        userRepository.save(user);

        // Force re-login on all other devices
        refreshTokenRepository.deleteByUserId(new ObjectId(user.getId()));
    }
}
